### DISCUSSION TITLE
Help with facial authentication.

### DISCUSSION THREAD

**Discussion by David-Braz** [community/community#143114] [GitHub Education]
### Select Topic Area

Question

### Body

Hello, I would like some help with a college project. I'm working on a project that aims to be a facial authenticator.
In this project, I'm using Java as the programming language, IntelliJ as the IDE, and OpenCV as the library.

I've been studying OpenCV and have already managed to do quite a bit with it, like converting images to grayscale, calculating histograms, and facial detection. However, I’m having serious doubts about how to authenticate. Specifically, how can I identify who is appearing on the webcam, and whether or not they have permission to access the system?

**Thank you for your attention; any help would be greatly appreciated.**

**Comment by DhareshP** [comment]
if it's a facial authenticator, then it means that you already have the data somewhere to verify the face right? so, simply, you train the model to trace the points or match the points with real time data, to do so you can collect a dataset of labeled images for each user and use face detection to identify faces in webcam frames, then extract features like using LBPH or a deep learning model then you train a classifier to recognize these features and during authentication, compare the live face against your trained model and set a confidence threshold for access and finally you grant or deny access based on whether the match meets your criteria

**Reply by David-Braz** [reply]
Thank you very much, this will help me a lot, now I have another question, could you explain a little better about the part of training my own model with the data that I already have today, I tried to look for a tutorial or something similar, but I saw that There is a lot of Python content, which ends up making it a little difficult since I'm using Java. Thanks again for the direction. @DhareshP

**Reply by DhareshP** [reply]
As you are done with the classification (labelling the images) also the more the images for a labelled user, the better will be the accuracy, and preprocessing the images using openCV (grayscale, resizing, crop, histograms).
which method of openCV are you using? by your question i assume it's LPBH, as you're generating histograms, so you're also done with feature extraction, now you only need to train your model, first prepare the data by storing the labels and images in arraylists, use openCV's class LBPFaceRecognizer , use LBPFaceRecognizer.create() to initialize the model. 
Now to train the recognizer with you data, simply pass the recognizer.train(imgs, Converters.vector_int_to_Mat(labels)); , it will convert the list of integer in a format that recognizer accepts. 
use the recognizer. predict() method, for returning user Id and accuracy.
for real time authentication: 
capture the real time data from cam, 
preprocess(crop, resize, grayscale)
use predict() from trained LBPHFaceRecognizer model to identify face
it will return the predicted label that is the (label id/ user id) and a score of accuracy
now you just have to grant the access as per predictedLabel and accuracy, you can use 

int authorizedUserId = 1;  //user id that has been predicted
double threshold = 50.0;   // accuracy 
if (accuracy[0] < threshold && predictedLabel[0] == authorizedUserID) {
    sout("access granted to"+ predictedLabel[0]);
} else {
    sout("access denied");
}
done, your model is ready

I'm glad I could help, if you have more questions on code or explanation please feel free to ask
